contracts/javascore/modules/ics20app/src/main/java/ibc/ics20app/ICS20Transfer.java

```diff
@@ -0,0 +1,214 @@
+    }
+
+    @External(readonly = true)
+    public String getDestinationChannel(String channelId) {
+        return destinationChannel.get(channelId);
+    }
+
+    @External
+    public byte[] onRecvPacket(byte[] packet, Address relayer) {
+        onlyIBC();
+        Packet packetDb = Packet.decode(packet);
+        ICS20Lib.PacketData data = ICS20Lib.unmarshalJSON(packetDb.getData());
+        boolean success = _decodeReceiver(data.receiver);
+        if (!success) {
+            return ICS20Lib.FAILED_ACKNOWLEDGEMENT_JSON;
+        }
+        Address receiver = Address.fromString(data.receiver);
+
+        byte[] denomPrefix = getDenomPrefix(packetDb.getSourcePort(), packetDb.getSourceChannel());
+        byte[] denom = data.denom.getBytes();
+
+        if (denom.length >= denomPrefix.length && Ops.hasPrefix(denom, denomPrefix)) {
+            byte[] unprefixedDenom = Arrays.copyOfRange(denom, denomPrefix.length, denom.length);
+            String unprefixedDenomString = new String(unprefixedDenom);
+            if (unprefixedDenomString.equals("icx")){
+                success = _transferICX(receiver, data.amount);
+            }
+            else {
+                success = _transferFrom(getEscrowAddress(packetDb.getDestinationChannel()), receiver, unprefixedDenomString, data.amount);
+            }
+        } else {
+            if (ICS20Lib.isEscapeNeededString(denom)) {
+                success = false;
+            } else {
+                denom = StringUtil.encodePacked(packetDb.getDestinationPort(), "/", packetDb.getDestinationChannel(), "/", data.denom);
+                String denomText = new String(denom);
+                success = _mint(receiver, denomText, data.amount);
+            }
+        }
+
+        if (success) {
+            return ICS20Lib.SUCCESSFUL_ACKNOWLEDGEMENT_JSON;
+        } else {
+            return ICS20Lib.FAILED_ACKNOWLEDGEMENT_JSON;
+        }
+    }
+
+
+    @External
+    public void onAcknowledgementPacket(byte[] packet, byte[] acknowledgement, Address relayer) {
+        onlyIBC();
+        Packet packetDb = Packet.decode(packet);
+        if (!acknowledgement.equals(ICS20Lib.SUCCESSFUL_ACKNOWLEDGEMENT_JSON)) {
+            refundTokens(ICS20Lib.unmarshalJSON(packetDb.getData()), packetDb.getSourcePort(), packetDb.getSourceChannel());
+        }
+    }
+
+    @External
+    public void onChanOpenInit(int order, String[] connectionHops, String portId, String channelId,
+                               byte[] counterpartyPb, String version) {
+        onlyIBC();
+        Context.require(order == Channel.Order.ORDER_UNORDERED, "must be unordered");
+        Context.require(version.equals(ICS20_VERSION), "version should be same with ICS20_VERSION");
+        Channel.Counterparty counterparty = Channel.Counterparty.decode(counterpartyPb);
+        destinationPort.set(channelId, counterparty.getPortId());
+        channelEscrowAddresses.set(channelId, Context.getAddress());
```

Comment:
Could this logic not be completely omitted, is there any scenario where escrow won't be this contract?